tools/syz-declextract/syz-declextract.cpp

```diff
@@ -64,6 +106,230 @@ bool beginsWith(const std::string_view &str, const std::string_view begin) {
+      return getFieldType(llvm::dyn_cast<BTFTagAttributedType>(field)->desugar(), context, fieldName);
+    case clang::Type::Elaborated:
+      return getFieldType(llvm::dyn_cast<ElaboratedType>(field)->desugar(), context, fieldName, parent, isSyscallParam);
+    case clang::Type::Enum: {
+      const auto &enumDecl = llvm::dyn_cast<EnumType>(field)->getDecl();
+      auto name = enumDecl->getNameAsString();
+      flags.push_back(name);
+      includes.push_back(std::filesystem::relative(SM->getFilename(enumDecl->getSourceRange().getBegin()).str()));
+      const char *sep = " = ";
+      for (const auto &enumerator : enumDecl->enumerators()) {
+        flags.back() += sep + enumerator->getNameAsString();
+        sep = ", ";
+      }
+      return "flags[" + name + "]";
+    }
+    case clang::Type::FunctionProto:
+      return makePtr("in", "int8");
+    default:
+      field->dump();
+      fprintf(stderr, "Unhandled field type %s\n", field->getTypeClassName());
+      exit(1);
+    }
+  }
+
+  std::string extractRecord(const RecordDecl *recordDecl, ASTContext *context, const std::string &backupName = "") {
+    recordDecl = recordDecl->getDefinition();
+    if (!recordDecl) { // When the definition is in a different translation unit.
+      return "intptr";
+    }
+    const auto &name = (recordDecl->getNameAsString().empty() ? backupName : recordDecl->getNameAsString());
+    const auto &recordName = name + "$auto_record";
+    if (visitedRecord.find(name) != visitedRecord.end()) { // Don't extract the same record twice.
+      return recordName;
+    }
+    visitedRecord[name] = extractedRecords.size();
+    extractedRecords.resize(extractedRecords.size() + 1);
+    bool isVarlen = false;
+    std::vector<StructMember> members;
+    for (const auto &field : recordDecl->fields()) {
+      std::string fieldName;
+      if (field->getName().empty()) {
+        fieldName = name + "_" + std::to_string(field->getFieldIndex());
+      } else if (field->isAnonymousStructOrUnion()) {
+        fieldName = name;
+      } else {
+        fieldName = field->getNameAsString();
+      }
+      const std::string &parentName = field->isAnonymousStructOrUnion() ? "" : name;
+      std::string fieldType = getFieldType(field->getType(), context, fieldName, parentName);
+      if (field->isUnnamedBitField()) { // pad
+        fieldType = makeConst(fieldType);
+      } else if (field->isBitField()) {
+        fieldType = fieldType + ":" + std::to_string(field->getBitWidthValue(*context));
+      }
+      if (fieldType.empty()) {
+        continue;
+      }
+      isVarlen |= isFieldVarlen(field->getType()) || (visitedRecord.find(fieldName) != visitedRecord.end() &&
+                                                      extractedRecords[visitedRecord[fieldName]].isVarlen);
+      members.push_back({fieldType, fieldName, getCountedBy(field)});
+    }
+    if (members.empty()) { // Empty structs are not allowed in Syzlang.
+      return "";
```

Comment:
It's very hard to understand what happens next with such types. Empty type "" is handled differently in different places, e.g. const handling just replaces it with intptr, which would be wrong in this case.

Is the only use case for this is recursive structs and triggering this branch above?
```
      if (fieldType.empty()) {
        continue;
      }
```

If yes, then a better way to do it would be to declare:

```
const std::string emptyStructType = "empty struct";
```
return `emptyStructType` here and do:
```
      if (fieldType == emptyStructType) {
        continue;
      }
```

It will make it possible to easily understand how exactly this special value is handled, and prevent it's accidental use in other places (if it will be printed to descriptions, then descriptions won't compile).

Reply:
Yes. That's the only use case.